node/src/components/storage.rs

```diff
@@ -1,805 +1,442 @@
+            .saturating_add(config.max_deploy_metadata_store_size);
+
+        // Creates the environment and databases.
+        let env = Environment::new()
+            .set_flags(
+                EnvironmentFlags::NO_SUB_DIR
+                    | EnvironmentFlags::NO_TLS
+                    | EnvironmentFlags::WRITE_MAP,
+            )
+            .set_max_readers(MAX_TRANSACTIONS)
+            .set_max_dbs(4)
+            .set_map_size(total_size)
+            .open(&root.join("storage.lmdb"))
+            .map_err(Error::LmdbInit)?;
+
+        let block_db = env
+            .create_db(Some("blocks"), DatabaseFlags::empty())
+            .map_err(Error::LmdbInit)?;
+        let deploy_db = env
+            .create_db(Some("deploys"), DatabaseFlags::empty())
+            .map_err(Error::LmdbInit)?;
+        let deploy_metadata_db = env
+            .create_db(Some("deploy_metadata"), DatabaseFlags::empty())
+            .map_err(Error::LmdbInit)?;
+
+        // Load chainspec from file. A corrupt chainspec will lead to a panic.
+        let chainspec_cache = if let Ok(raw) = fs::read(root.join(CHAINSPEC_CACHE_FILENAME)) {
+            Some(deser(&raw))
+        } else {
+            None
+        };
 
-    fn get_deploy_headers(
-        &self,
-        deploy_hashes: DeployHashes<Self>,
-        responder: Responder<DeployHeaderResults<Self>>,
-    ) -> Effects<Event<Self>>
-    where
-        Self: Sized,
-    {
-        let deploy_store = self.deploy_store();
-        async move {
-            let results = task::spawn_blocking(move || deploy_store.get_headers(deploy_hashes))
-                .await
-                .expect("should run")
-                .into_iter()
-                .map(|result| {
-                    result.unwrap_or_else(|error| panic!("failed to get deploy header: {}", error))
-                })
-                .collect();
-            responder.respond(results).await
+        // We now need to restore the block-height index. Log messages allow timing here.
+        info!("reindexing block store");
+        let mut block_height_index = BTreeMap::new();
+        let block_tx = env.ro_transaction();
+        let mut cursor = block_tx
+            .open_ro_cursor(block_db)
+            .expect("could not create read-only cursor on block store");
+
+        // Note: `iter_start` has an undocument panic if called on an empty database. We rely on
+        //       the iterator being at the start when created.
+        for (raw_key, raw_val) in cursor.iter() {
+            let block: Block = deser(raw_val);
+            // We use the opportunity for a small integrity check.
+            assert_eq!(
+                raw_key,
+                block.hash().as_ref(),
+                "found corrupt block in database"
+            );
+            let header = block.header();
+            if let Some(duplicate) = block_height_index.insert(header.height(), *block.hash()) {
+                warn!(
```

Comment:
Should be a panic probably

Reply:
Actually I think we're looking at this wrong. We're opening ourselves up to attacks this way.

We should tag each block with a chain identifier, or only accept blocks that match our chainspec/genesis and refuse to add those we don't know.

Otherwise someone could simply send us a block at a specific height from a different chain, and we'd panic.

Reply:
We do that in other places that accept blocks. Storage should not be concerned about it.

Reply:
Hmm, not sure I agree, but I am making it `fatal` for now.